## About the sheet: 

This sheet is being used to prepare the data which would be used for creating the RAG system and then finally pushing the data to S3 bucket.

In [1]:
import boto3
import pandas as pd
import numpy as np
import json
import pandas as pd
import dotenv
import os

In [4]:
s3 = boto3.client('s3')

response = s3.list_buckets()
for bucket in response['Buckets']:
    print(f'Bucket Name: {bucket["Name"]}')


Bucket Name: menudatabucket


In [5]:
# Fetching data from S3 bucket and saving to local
bucket_name = 'menudatabucket'
file_key = 'Sample_Ingredients_File.csv' 
local_filename = 'Sample_Ingredients_File.csv'

s3.download_file(bucket_name, file_key, local_filename)
print(f"File {file_key} downloaded as {local_filename}")


File Sample_Ingredients_File.csv downloaded as Sample_Ingredients_File.csv


In [408]:
Ingredients = pd.read_csv(f"Sample_Ingredients_File.csv")
print(Ingredients.shape)

(52696, 16)


In [23]:
Ingredients["item_id"].nunique()

10571

In [18]:
Ingredients["menu_category"].nunique()

998

In [15]:
Ingredients["restaurant_name"].nunique()

150

We have information on

> 52696 items.

> 150 restaurants.

> 998 unique menu categories. Here we can see: particular day's specials(Tuesday 11-9pm), drink categories, breakfast, etc.

> 10571 unique items.

> Restaurants only from san francisco and 1 pincode.

### Interrpretation of data:

restaurant_name: Name of the restaurant.

Example: '20 spot' (a restaurant in San Francisco).
menu_category: Category of the menu item.

Example: 'no proof' (likely indicating a non-alcoholic or low-alcohol drink).
item_id: Unique identifier for the menu item.

Example: 24932147 (a unique number assigned to the menu item).
menu_item: Name of the menu item.

Example: '"amaro" spritz' (a type of cocktail or drink).
menu_description: Brief description of the menu item, often listing key ingredients.

Example: 'pathfinder amaro, tonic' (describing the components of the drink).
ingredient_name: Key ingredient(s) used in the menu item.

Example: 'pathfinder amaro' (a type of non-alcoholic spirit).
confidence: Confidence score (possibly from an ML model) indicating the certainty of ingredient classification.

Example: 0.95 (95% confidence that the ingredient identification is correct).
categories: Cuisine type(s) or restaurant classification, separated by |.

Example: 'New American|Wine Bars' (restaurant serves New American cuisine and has a wine bar).


> No proof in menu_category refers to little to no alcohol for that item.

In [63]:
for col in Ingredients.columns:
    print(f"{col.strip()} has null values: {Ingredients[col].isnull().sum()}")

restaurant_name has null values: 0
menu_category has null values: 13
item_id has null values: 0
menu_item has null values: 0
menu_description has null values: 13793
ingredient_name has null values: 42
confidence has null values: 0
categories has null values: 140
address1 has null values: 140
city has null values: 140
zip_code has null values: 140
country has null values: 140
state has null values: 140
rating has null values: 140
review_count has null values: 140
price has null values: 8608


**Handling Missing Values:**

Solution 1: We can fill out menu_description and menu_category using a large language model. We can provide context using other details. 

Solution 2: Another potential solution could be, integrating with some food delivery app and use their API if its accessible and then get the missing values.

For now, lets work with Solution 1. I am going to be using meta llama for the same.




## Setting up Bedrock

In [28]:
load_dotenv("Credentials")

aws_access_key_id = os.getenv("aws_access_key_id")
aws_secret_access_key = os.getenv("aws_secret_access_key")


'AKIAUMYCIE27A72XFOAK'

In [259]:
bedrock_client = boto3.client(
    service_name = "bedrock-runtime",
    region_name="us-east-2",
    aws_access_key_id = aws_access_key_id, 
    aws_secret_access_key = aws_secret_access_key
)

## Getting null values in all the columns

In [270]:
rows_with_null_vals = Ingredients[
    Ingredients[['menu_category', 'menu_item', 'menu_description', 'ingredient_name', 'categories']].isnull().any(axis=1)]
rows_with_null_vals["menu_description"].isnull().sum()



13793

### Filling out missing values in 'menu_description' columns

In [271]:
def generate_prompt_description(row):
    menu_category = row.get("menu_category", "")
    menu_item = row.get("menu_item", "")
    ingredient_name = row.get("ingredient_name", "")
    categories = row.get("categories", "")

    prompt = f"""You are a professional menu writer known for crafting **irresistible, mouthwatering** descriptions that captivate customers. Your task is to create a **vivid, enticing** menu description based on the given details.

### Instructions:
- **Always generate a description**—never leave it empty.
- Use the **menu item name, ingredient name, menu category, and categories** as inspiration.
- If an ingredient is provided, make it the star of the dish.
- **For food:** Use **rich, sensory phrases** to describe flavor, texture, and experience.
- **For drinks:** Keep it **short, crisp, and refreshing**, focusing on key ingredients and taste.
- **No explanations—only output the final description.**

### Examples:

#### **Drinks (Short & Crisp)**
1. **Menu Category:** No Proof  
   **Menu Item:** Pathfinder Tonic  
   **Ingredient:** Pathfinder Amaro  
   **Output:** Herbal, citrusy, and refreshingly bitter with a bright tonic finish.  

2. **Menu Category:** Cocktails  
   **Menu Item:** Smoked Old Fashioned  
   **Ingredient:** Bourbon  
   **Output:** Smooth bourbon, smoked oak, orange essence, and a hint of spice.  

3. **Menu Category:** Refreshers  
   **Menu Item:** Cucumber Cooler  
   **Ingredient:** Cucumber  
   **Output:** Crisp cucumber, fresh lime, and a cooling mint finish.  

#### **Food (Rich & Evocative)**
4. **Menu Category:** Dessert  
   **Menu Item:** Blueberry Dessert  
   **Ingredient:** Blueberry  
   **Output:** A luscious medley of plump blueberries, silky chantilly cream, and a whisper of fresh mint.  

5. **Menu Category:** Mains  
   **Menu Item:** Braised Oxtails  
   **Ingredient:** Braised Oxtails  
   **Output:** Slow-braised oxtails in a rich sungold tomato sauce, with blistered shishito peppers and fragrant herbs.  

6. **Menu Category:** Snacks  
   **Menu Item:** Marinated Olives  
   **Ingredient:** Castelvetrano Olives  
   **Output:** Buttery Castelvetrano olives kissed with citrus zest, rosemary, and a hint of chili heat.  

### Generate a description for:
- **Menu Category:** {menu_category}  
- **Menu Item:** {menu_item}  
- **Ingredient:** {ingredient_name}  
- **Categories:** {categories}  

**Output only the final description. No explanations or extra text.**"""

    return prompt




# Temperature can be kept higher here, creativity would be good for this task
def get_description_from_bedrock(prompt):
    model_id = "meta.llama3-3-70b-instruct-v1:0"
    payload = {
        "prompt": prompt,
        "temperature": 0.7
    }
    
    try:
        response = bedrock_client.invoke_model(
            modelId=model_id,
            body=json.dumps(payload),
            contentType="application/json",
            accept="application/json"
        )
        result = json.loads(response["body"].read().decode("utf-8"))
        description = result.get("generation", "").strip()
        
        description = description.split('\n')[0]
        description = description.rstrip('.') 
        description = description.lower() 
        
        return description
    except Exception as e:
        print(f"Error invoking model: {e}")
        return "default preparation"  

desc_cnt = 0


for index, row in rows_with_null_vals.iterrows():
    if pd.isnull(row["menu_description"]) and desc_cnt <=50:
        menu_category = row.get("menu_category", "")
        menu_item = row.get("menu_item", "")
        ingredient_name = row.get("ingredient_name", "")
        categories = row.get("categories", "")

        print(f"  Category: {menu_category}")
        print(f"  Item: {menu_item}")
        print(f"  Ingredient: {ingredient_name}")
        print(f"  Categories: {categories}")

        prompt = generate_prompt_description(row)
        generated_description = get_description_from_bedrock(prompt)

        rows_with_null_vals.at[index, "menu_description"] = generated_description
        print(f"  Generated Description: {generated_description}\n")
        desc_cnt+=1
        print(desc_cnt, "done")
        print()


  Category: pet-nat & sparkling wine
  Item: athenais de beru, ‘love joy’, chardonnay
  Ingredient: athenais de beru, 'love joy', chardonnay
  Categories: New American|Wine Bars
  Generated Description: fresh, lively, and effervescent, with notes of green apple, citrus zest, and a hint of floral sweetness.  

1 done

  Category: pet-nat & sparkling wine
  Item: broc cellars, valdigué
  Ingredient: nan
  Categories: New American|Wine Bars
  Generated Description: sparkling valdigué with a lively, fruity nose and a crisp, refreshing finish.  

2 done

  Category: pet-nat & sparkling wine
  Item: cantina marilina, 'fedelie', moscato
  Ingredient: moscato
  Categories: New American|Wine Bars
  Generated Description: petite bubbles dance on the palate, as moscato's floral sweetness and crisp citrus notes unfold in this lively, sun-kissed italian sparkler.  

3 done

  Category: pet-nat & sparkling wine
  Item: casa coste piane, 'brichet', glera/verdiso
  Ingredient: glera
  Categories: New 

In [273]:
rows_with_null_vals["menu_description"].isnull().sum()

13742

## Filling Category Information

> On the basis of all categories, on a high level I can club them into these categories.
>
>  This is done so as to reduce the prompt length where we would have to pass a huge prompt. As we have almost 1000 categories.

In [279]:


# Allowed categories for strict matching
allowed_categories = {
    "Meals & Main Courses",
    "Breakfast & Brunch",
    "Appetizers & Sides",
    "Beverages",
    "Desserts & Sweets",
    "Specialty Menus & Offers",
    "Alcoholic Beverages & Wines",
    "International/Regional Dishes",
    "Healthy & Dietary Specific",
    "Seasonal & Limited Offerings",
    "Specialty Drinks",
    "Kids & Family Menus",
    "Snacks & Quick Bites",
    "Fresh & Organic Options",
    "Meals with Alcohol Pairings",
    "Unknown",
}

def generate_prompt_category(row):
    """Generate a prompt for categorization based on menu details."""
    menu_item = row.get("menu_item", "")
    menu_description = row.get("menu_description", "")
    ingredient_name = row.get("ingredient_name", "")

    prompt = f"""You are a food categorization system. Output ONLY the category name, no explanations, no additional text.

### RULES:
1. Output ONLY the category name.
2. NO explanations or extra text.
3. NO punctuation.
4. Choose from these categories ONLY:
{', '.join(allowed_categories)}

### Examples:

#### Correct Outputs:
- **Input:** Pancakes, fluffy with maple syrup  
  **Output:** Breakfast & Brunch  

- **Input:** Spaghetti Carbonara, creamy sauce with pancetta  
  **Output:** Meals & Main Courses  

- **Input:** Chocolate Lava Cake, rich molten chocolate center  
  **Output:** Desserts & Sweets  

- **Input:** Unknown dish with no clear category  
  **Output:** Unknown  

### Categorize this item:
- **Menu Item:** {menu_item}  
- **Menu Description:** {menu_description}  
- **Main Ingredient:** {ingredient_name}  

**Output only the category name.**"""

    return prompt

def get_category_from_bedrock(prompt):
    """Invoke the Bedrock model and retrieve the category."""
    model_id = "meta.llama3-3-70b-instruct-v1:0"
    payload = {
        "prompt": prompt,
        "temperature": 0.2  # Low temperature for consistency
    }
    
    try:
        response = bedrock_client.invoke_model(
            modelId=model_id,
            body=json.dumps(payload),
            contentType="application/json",
            accept="application/json"
        )
        result = json.loads(response["body"].read().decode("utf-8"))
        
        # Ensure category is correctly extracted
        category = result.get("generation", "").strip().split("\n")[0].strip().rstrip('.')

        # Return category only if valid, otherwise default to "Unknown"
        return category if category in allowed_categories else "Unknown"
        
    except Exception as e:
        print(f"Error invoking model: {e}")
        return "Unknown"

for index, row in rows_with_null_vals.iterrows():
    if pd.isnull(row["menu_category"]):
        
        menu_category = row.get("menu_category", "")
        menu_item = row.get("menu_item", "")
        ingredient_name = row.get("ingredient_name", "")
        categories = row.get("categories", "")

        print(f"Category: {menu_category}")
        print(f"Item: {menu_item}")
        print(f"Ingredient: {ingredient_name}")
        print(f"Categories: {categories}")

        prompt = generate_prompt_category(row)
        predicted_category = get_category_from_bedrock(prompt)
        rows_with_null_vals.at[index, "menu_category"] = predicted_category
        print(f"Predicted Category: {predicted_category}")
        print()


Category: nan
Item: ceviche tostadas
Ingredient: chicken
Categories: Burgers|Breakfast & Brunch|New American
Predicted Category: International/Regional Dishes

Category: nan
Item: ceviche tostadas
Ingredient: flounder
Categories: Burgers|Breakfast & Brunch|New American
Predicted Category: Appetizers & Sides

Category: nan
Item: ceviche tostadas
Ingredient: fried flounder
Categories: Burgers|Breakfast & Brunch|New American
Predicted Category: Appetizers & Sides

Category: nan
Item: ceviche tostadas
Ingredient: lemon juice
Categories: Burgers|Breakfast & Brunch|New American
Predicted Category: Appetizers & Sides

Category: nan
Item: ceviche tostadas
Ingredient: onions
Categories: Burgers|Breakfast & Brunch|New American
Predicted Category: Appetizers & Sides

Category: nan
Item: ceviche tostadas
Ingredient: peppers
Categories: Burgers|Breakfast & Brunch|New American
Predicted Category: Appetizers & Sides

Category: nan
Item: ceviche tostadas
Ingredient: salmon
Categories: Burgers|Breakfas

In [281]:
rows_with_null_vals["menu_category"].isnull().sum()

0

In [344]:
import re
import json
import pandas as pd

def generate_ingredient_prompt(row):
    """Generate a refined prompt for the LLM, passing the extracted ingredient for validation."""
    menu_item = row.get("menu_item", "")
    menu_description = row.get("menu_description", "")
    menu_category = row.get("menu_category", "")
    categories = row.get("categories", "")
    
    prompt = f"""You are a very intelligent AI assistant that extracts key ingredients from food items. 
Your task is to see the menu_category, menu_item, and categories, accordingly give the key ingredients for this food item.
### Examples:
#### Input:  
**Menu Category:** Pizza  
**Menu Item:** Margherita Pizza  
**Menu Description:** Fresh basil, tomato sauce, mozzarella  
**Extracted Ingredient:** basil, tomato, sauce  
**Categories:** Italian, Vegetarian  
#### Output:  
mozzarella, basil, tomato  

#### Input:  
**Menu Category:** Sandwiches  
**Menu Item:** BBQ Pulled Pork  
**Menu Description:** Slow-cooked pork in smoky BBQ sauce  
**Extracted Ingredient:** pork, smoky, BBQ  
**Categories:** American, Meat  
#### Corrected Output:  
pulled pork, BBQ sauce  

---
### Your task:  
Given the following Menu Item, Menu Description, Menu Category and categories, give a list that contains only the key ingredients for this food item.  
- **Menu Item:** {menu_item}  
- **Menu Description:** {menu_description}  
- **Menu Category:** {menu_category}  
- **Categories:** {categories} 

**Output Instructions**
- Give only the ingredient list, separated by commas.
- Limit to **4-5 key ingredients**.  
- No explanations, extra words, or formatting—just the list.

**Output only the ingredients.**

"""
    return prompt


def clean_output(ingredient_str):
    """Clean the output to ensure only 4-5 ingredients are included."""
    # Split the ingredients by commas
    ingredients = ingredient_str.split(",")
    
    # Limit to the first 4 ingredients using regex and strip extra spaces
    ingredients = [ingredient.strip() for ingredient in ingredients[:4]]
    
    # Join the ingredients back together with commas
    return ", ".join(ingredients)


def get_ingredient_from_bedrock(prompt):
    """Invoke the Bedrock model to get an ingredient name."""
    model_id = "meta.llama3-3-70b-instruct-v1:0"
    payload = {
        "prompt": prompt,
        "temperature": 0.5
    }
    
    try:
        response = bedrock_client.invoke_model(
            modelId=model_id,
            body=json.dumps(payload),
            contentType="application/json",
            accept="application/json"
        )
        result = json.loads(response["body"].read().decode("utf-8"))
        
        ingredient = result.get("generation", "").strip().split("\n")[0].strip().rstrip('.')
        return clean_output(ingredient) if ingredient else "Unknown"
        
    except Exception as e:
        print(f"Error invoking model: {e}")
        return "Unknown"

cnt = 0
for index, row in rows_with_null_vals.iterrows():
    if pd.isnull(row["ingredient_name"]) and cnt <=50:
        
        # Generate refined LLM prompt and get cleaned ingredients
        prompt = generate_ingredient_prompt(row)
        final_ingredient = get_ingredient_from_bedrock(prompt)
        menu_category = row.get("menu_category", "")
        menu_item = row.get("menu_item", "")
        categories = row.get("categories", "")
        print(f"Category: {menu_category}")
        print(f"Item: {menu_item}")
        print(f"Categories: {categories}")
        print(f"Predicted ingredients: {final_ingredient}")
        rows_with_null_vals.at[index, "ingredient_name"] = final_ingredient  
        cnt += 1
        print(cnt, "done\n")


Category: pet-nat & sparkling wine
Item: broc cellars, valdigué
Categories: New American|Wine Bars
Predicted ingredients: valdigué, grapes, yeast, sugar
1 done

Category: sweet dessert wines
Item: loupiac, chateau dauphine rondeillon, 2009
Categories: New American
Predicted ingredients: grapes, sugar, yeast, water
2 done

Category: dessert - fortified wines
Item: ruby port, lbv, faria vinhos, 2015
Categories: New American
Predicted ingredients: ruby, port, grapes, wine
3 done

Category: dessert - fortified wines
Item: tawny port, taylor fladgate, 2017
Categories: New American
Predicted ingredients: tawny port, grapes, wine, taylor fladgate
4 done

Category: pu-erh
Item: 1998 ripe shu pu-erh tea cake
Categories: Coffee & Tea|Juice Bars & Smoothies|Vegan
Predicted ingredients: pu-erh tea leaves, tea, pu-erh, tea cake
5 done

Category: pu-erh
Item: 2013 ripe pu-erh tea and jiaogulan
Categories: Coffee & Tea|Juice Bars & Smoothies|Vegan
Predicted ingredients: pu-erh tea, jiaogulan  tea lea

In [345]:
rows_with_null_vals.to_csv("rows_with_null_vals.csv")

In [393]:
rows_with_null_vals = pd.read_csv("rows_with_null_vals.csv").drop(["Unnamed: 0"], axis = 1)


> For demo purpose, we have filled the data here. Similar approach can be taken to fill categories also.

In [362]:
Ingredients

restaurant_name             menu_category   item_id  \
0             20 spot                  no proof  24932147   
1             20 spot                  no proof  24932146   
2             20 spot                  no proof  24932145   
3             20 spot                  no proof  24932145   
4             20 spot  pet-nat & sparkling wine  24932165   
...               ...                       ...       ...   
52691       yucatasia       maya especialidades  24398826   
52692       yucatasia       maya especialidades  24398828   
52693       yucatasia       maya especialidades  24398828   
52694       yucatasia       maya especialidades  24398828   
52695       yucatasia       maya especialidades  24398828   

                                      menu_item  \
0                                "amaro" spritz   
1                                 "gin & tonic"   
2                                 amalfi spritz   
3                                 amalfi spritz   
4      athenais de beru, ‘love joy’, chardonnay   
...                                         ...   
52691                         49. cabeza de res   
52692                                 50. pibes   
52693                                 50. pibes   
52694                                 50. pibes   
52695                                 50. pibes   

                                        menu_description  \
0                                pathfinder amaro, tonic   
1                                                 lyre's   
2                                                 lyre's   
3                                                 lyre's   
4                                                    NaN   
...                                                  ...   
52691  beef tongue tossed in cabbage and radish serve...   
52692  baked corn masa stuffed with chicken and pork ...   
52693  baked corn masa stuffed with chicken and pork ...   
52694  baked corn masa stuffed with chicken and pork ...   
52695  baked corn masa stuffed with chicken and pork ...   

                                ingredient_name  confidence  \
0                              pathfinder amaro        0.95   
1                                           gin        0.80   
2                                 amalfi spritz        0.95   
3                                        lyre's        0.80   
4      athenais de beru, 'love joy', chardonnay        0.90   
...                                         ...         ...   
52691                                  tortilla        0.90   
52692                                   chicken        0.80   
52693                                 corn masa        0.90   
52694                                      pork        0.80   
52695                                     salsa        0.70   

                   categories         address1           city  zip_code  \
0      New American|Wine Bars     3565 20th St  San Francisco   94110.0   
1      New American|Wine Bars     3565 20th St  San Francisco   94110.0   
2      New American|Wine Bars     3565 20th St  San Francisco   94110.0   
3      New American|Wine Bars     3565 20th St  San Francisco   94110.0   
4      New American|Wine Bars     3565 20th St  San Francisco   94110.0   
...                       ...              ...            ...       ...   
52691                 Mexican  2164 Mission St  San Francisco   94110.0   
52692                 Mexican  2164 Mission St  San Francisco   94110.0   
52693                 Mexican  2164 Mission St  San Francisco   94110.0   
52694                 Mexican  2164 Mission St  San Francisco   94110.0   
52695                 Mexican  2164 Mission St  San Francisco   94110.0   

      country state  rating  review_count price  
0          US    CA     4.3         270.0    $$  
1          US    CA     4.3         270.0    $$  
2          US    CA     4.3         270.0    $$  
3          US    CA     4.3         270.0    $$  
4          US    CA     4.3    

In [365]:
Ingredients[Ingredients["address1"].isnull()]["restaurant_name"].unique()

array(['fort point valencia'], dtype=object)

In [366]:
Ingredients[Ingredients["zip_code"].isnull()]["restaurant_name"].unique()

array(['fort point valencia'], dtype=object)

> Only for fort point valencia we dont have address.

> We can use google maps api. However, we just have single missing entry here. That can be filled by manually searching.
>
> 742 Valencia St, San Francisco, CA 94110

In [399]:
restaurants_with_null_zip = rows_with_null_vals[Ingredients["zip_code"].isnull()]
unique_restaurants = restaurants_with_null_zip["restaurant_name"].unique()
rows_with_null_vals.loc[rows_with_null_vals["zip_code"].isnull(), "zip_code"] = 94110
rows_with_null_vals.loc[rows_with_null_vals["address1"].isnull(), "address1"] = "742 Valencia St"  
rows_with_null_vals.loc[rows_with_null_vals["city"].isnull(), "city"] = "San Francisco"  
rows_with_null_vals.loc[rows_with_null_vals["state"].isnull(), "state"] = "CA"  
rows_with_null_vals.loc[rows_with_null_vals["country"].isnull(), "country"] = "US"  

/tmp/ipykernel_7997/3178223702.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  restaurants_with_null_zip = rows_with_null_vals[Ingredients["zip_code"].isnull()]


In [400]:
rows_with_null_vals.to_csv("rows_with_null_vals.csv")

In [442]:
rows_with_null_vals = pd.read_csv("rows_with_null_vals.csv").drop(["Unnamed: 0"], axis = 1)


In [443]:
item_ids_to_remove = rows_with_null_vals["item_id"].unique()

Ingredients_clean = Ingredients[~Ingredients["item_id"].isin(item_ids_to_remove)]

Ingredients_imputed = pd.concat([Ingredients_clean, rows_with_null_vals], ignore_index=True)

for col in Ingredients_imputed:
    Ingredients_imputed[col].fillna("Unknown", inplace = True)


/tmp/ipykernel_7997/1678889898.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  Ingredients_imputed[col].fillna("Unknown", inplace = True)
/tmp/ipykernel_7997/1678889898.py:8: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'Unknown' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  Ingredients_imputed[col].fillna("Unknown", inplace = True)


In [448]:

Ingredients_imputed['menu_description'] = Ingredients_imputed['menu_description'].str.replace(r'[^\w\s]', '', regex=True).str.lower()
Ingredients_imputed['ingredient_name'] = Ingredients_imputed['ingredient_name'].str.replace(r'[^\w\s]', '', regex=True).str.lower()
Ingredients_imputed['restaurant_name'] = Ingredients_imputed['restaurant_name'].str.replace(r'[^\w\s]', '', regex=True).str.lower()
Ingredients_imputed['menu_item'] = Ingredients_imputed['menu_item'].str.replace(r'[^\w\s]', '', regex=True).str.lower()
Ingredients_imputed

Unnamed: 0 restaurant_name      menu_category   item_id  \
0               0         20 spot           no proof  24932147   
1               1         20 spot           no proof  24932146   
2               2         20 spot           no proof  24932145   
3               3         20 spot           no proof  24932145   
4               4         20 spot           no proof  24932150   
...           ...             ...                ...       ...   
52668       52668          yasmin  syrian quesadilla  26086027   
52669       52669          yasmin  syrian quesadilla  26086027   
52670       52670          yasmin  syrian quesadilla  26086027   
52671       52671          yasmin  syrian quesadilla  26086027   
52672       52672          yasmin  syrian quesadilla  26086027   

                       menu_item        menu_description   ingredient_name  \
0                   amaro spritz  pathfinder amaro tonic  pathfinder amaro   
1                     gin  tonic                   lyres               gin   
2                  amalfi spritz                   lyres     amalfi spritz   
3                  amalfi spritz                   lyres             lyres   
4                   blood orange          san pellegrino      blood orange   
...                          ...                     ...               ...   
52668  syrian chicken quesadilla                     NaN         olive oil   
52669  syrian chicken quesadilla                     NaN            onions   
52670  syrian chicken quesadilla                     NaN           parsley   
52671  syrian chicken quesadilla                     NaN        pita bread   
52672  syrian chicken quesadilla                     NaN          tomatoes   

       confidence              categories         address1           city  \
0            0.95  New American|Wine Bars     3565 20th St  San Francisco   
1            0.80  New American|Wine Bars     3565 20th St  San Francisco   
2            0.95  New American|Wine Bars     3565 20th St  San Francisco   
3            0.80  New American|Wine Bars     3565 20th St  San Francisco   
4            0.90  New American|Wine Bars     3565 20th St  San Francisco   
...           ...                     ...              ...            ...   
52668        0.50           Mediterranean  799 Valencia St  San Francisco   
52669        0.65           Mediterranean  799 Valencia St  San Francisco   
52670        0.55           Mediterranean  799 Valencia St  San Francisco   
52671        0.75           Mediterranean  799 Valencia St  San Francisco   
52672        0.60           Mediterranean  799 Valencia St  San Francisco   

       zip_code country state  rating  review_count price  
0       94110.0      US    CA     4.3         270.0    $$  
1       94110.0      US    CA     4.3         270.0    $$  
2       94110.0      US    CA     4.3         270.0    $$  
3       94110.0      US    CA     4.3         270.0    $$  
4       94110.0      US    CA     4.3         270.0    $$  
...         ...     ...   ...     ...           ...   ...  
52668   94110.0      US    CA     4.2         160.0    $$  
52669   94110.0      US    CA     4.2         160.0    $$  
52670   94110.0      US    CA     4.2         160.0    $$  
52671   94110.0      US    CA     4.2         160.0    $$  
52672   94110.0      US    CA     4.2         160.0    $$  

[52673 rows x 17 columns]

In [449]:
Ingredients_imputed.to_csv("Ingredients_imputed.csv")

In [450]:
local_filename = "Ingredients_imputed.csv"
s3.upload_file(local_filename, bucket_name, file_key)
print(f"File {local_filename} uploaded to bucket {bucket_name} with key {file_key}")


File Ingredients_imputed.csv uploaded to bucket menudatabucket with key Sample_Ingredients_File.csv


> Cleaned the data.

> improved the descriptions by generating comprehensive descriptions on which embeddings would be created.

> Finally pushed the data to S3 bucket.